packages/replayio/src/utils/createPromiseQueue.ts

```diff
@@ -0,0 +1,91 @@
+import { Deferred, createDeferred } from "./createDeferred";
+
+interface PromiseQueueRoot {
+  add: <T>(action: () => Promise<T>) => Promise<T>;
+}
+
+interface PromiseQueue extends PromiseQueueRoot {
+  fork: () => PromiseQueue;
+  waitUntilIdle: () => Promise<undefined>;
+}
+
+function createGroup(root: PromiseQueueRoot): PromiseQueue {
+  let pendingCount = 0;
+  let idleDeferred: Deferred<undefined> | undefined;
+
+  return {
+    add: <T>(action: () => Promise<T>) => {
+      pendingCount++;
+      const finalize = <T>(result: T) => {
+        pendingCount--;
+        if (!pendingCount) {
+          idleDeferred?.resolve();
+          idleDeferred = undefined;
+        }
+        return result;
+      };
+      return root.add(() => action().then(finalize, finalize));
```

Comment:
Returning `result` for the 2nd callback will cause rejected Promises to be treated as successful Promises (that returned an error instead of throwing one). That seems unintentional?

Reply:
the `result` there is a misnomer - it's actually either a result or an error (since the same function is passed as both `onfulfilled` and `onrejected`). I'll rename it to `resultOrError` to make it clear

Reply:
Wouldn't we still be swallowing errors since (a) this method always calls `idleDeferred?.resolve()` (without checking for failure) and (b) it always wraps the passed in `action` (Promise) and returns the result instead of throwing it (which would also cause a rejected Promise to become resolved by the wrapper)

Reply:
Calling `idleDeferred?.resolve()` is intentional here. This queue isn't responsible for handling errors anyhow - it's up to each queued job.

You are right about throwing, a rejected promise is not the same as one resolved with an error 😅

Reply:
> Calling idleDeferred?.resolve() is intentional here. This queue isn't responsible for handling errors anyhow - it's up to each queued job.

That makes sense.

Tests will help clarify the expected behavior I guess.